### PR TITLE
refactor(clustering): small improvements about incremental sync

### PR DESCRIPTION
**Pull request by chronolaw**
### Summary

KAG-5768
KAG-5776
KAG-5780
KAG-5788

### Checklist

- [ ] The Pull Request has tests
- [ ] A changelog file has been created under `changelog/unreleased/kong` or `skip-changelog` label added on PR if changelog is unnecessary. [README.md](https://github.com/Kong/gateway-changelog/blob/main/README.md)
- [ ] There is a user-facing docs PR against https://github.com/Kong/docs.konghq.com - PUT DOCS PR HERE

### Issue reference

Fix #_[issue number]_

### CODE REVIEW

File: kong/clustering/services/sync/hooks.lua [hunk 1]
```diff
@@ -67,11 +67,8 @@ function _M:notify_all_nodes()
     local res, err = kong.rpc:call(node, "kong.sync.v2.notify_new_version", msg)
     if not res then
       if not err:find("requested capability does not exist", nil, true) then
-        ngx_log(ngx_ERR, "unable to notify new version: ", err)
+        ngx_log(ngx_ERR, "unable to notify ", node, " new version: ", err)
       end
-
-    else
-      ngx_log(ngx_DEBUG, "notified ", node, " ", latest_version)
```
Comment (StarlightIbuki) [inline kong/clustering/services/sync/hooks.lua:74]: I think we still need the log for debugging. It will be hard to diagnose if we log nothing when succeeds

File: kong/clustering/services/sync/rpc.lua [hunk 2]
```diff
@@ -206,7 +203,7 @@ local function do_sync()
   local deltas = ns_delta.deltas
 
   if isempty(deltas) then
-    ngx_log(ngx_DEBUG, "no delta to sync")
```
Comment (StarlightIbuki) [inline kong/clustering/services/sync/rpc.lua:209]: So is here. Or we can log the number of deltas

File: kong/clustering/services/sync/hooks.lua (hunk 1 above)
Comment (chronolaw) [reply kong/clustering/services/sync/hooks.lua:74]: See comment: https://github.com/Kong/kong-ee/pull/10460#discussion_r1831761668

Comment (chronolaw) [reply kong/clustering/services/sync/hooks.lua:74]: Let's remove it now, If we do need it in the practice we could add them back.

File: kong/clustering/services/sync/rpc.lua [hunk 3]
```diff
@@ -88,7 +88,7 @@ function _M:init_cp(manager)
       sync_status = CLUSTERING_SYNC_STATUS.NORMAL,
       config_hash = fmt("%032d", default_namespace_version),
       rpc_capabilities = rpc_peers and rpc_peers[node_id] or {},
-    }, { ttl = purge_delay })
+    }, { ttl = purge_delay, no_broadcast_crud_event = true })
```
Comment (chronolaw) [inline kong/clustering/services/sync/rpc.lua:91]: Perhaps we should not have this `no_broadcast_crud_event` flag

File: kong-3.9.0-0.rockspec [hunk 4]
```diff
@@ -322,7 +322,7 @@ build = {
     ["kong.db.migrations.core.021_340_to_350"] = "kong/db/migrations/core/021_340_to_350.lua",
     ["kong.db.migrations.core.022_350_to_360"] = "kong/db/migrations/core/022_350_to_360.lua",
     ["kong.db.migrations.core.023_360_to_370"] = "kong/db/migrations/core/023_360_to_370.lua",
-    ["kong.db.migrations.core.024_370_to_380"] = "kong/db/migrations/core/024_370_to_380.lua",
```
Comment (StarlightIbuki) [inline kong-3.9.0-0.rockspec:325]: Is this intended? We do not need `370_to_380`?

File: kong/clustering/services/sync/rpc.lua [hunk 5]
```diff
@@ -88,7 +88,7 @@ function _M:init_cp(manager)
       sync_status = CLUSTERING_SYNC_STATUS.NORMAL,
       config_hash = fmt("%032d", default_namespace_version),
       rpc_capabilities = rpc_peers and rpc_peers[node_id] or {},
-    }, { ttl = purge_delay })
+    }, { ttl = purge_delay, })
```
Comment (StarlightIbuki) [inline kong/clustering/services/sync/rpc.lua:91]: Unnecessary change

Comment (chronolaw) [reply kong/clustering/services/sync/rpc.lua:91]: style requirement.

File: kong-3.9.0-0.rockspec (hunk 4 above)
Comment (chronolaw) [reply kong-3.9.0-0.rockspec:325]: Yes, as we talking offline meetin.

File: kong/clustering/services/sync/hooks.lua (hunk 1 above)
Comment (StarlightIbuki) [reply kong/clustering/services/sync/hooks.lua:74]: @chronolaw With the level debug it's almost free in runtime. Any specific reason we are removing them?

**Review by chobits — APPROVED** [review]

File: kong/clustering/services/sync/rpc.lua (hunk 3 above)
Comment (chobits) [reply kong/clustering/services/sync/rpc.lua]: @chronolaw has rolled back this fix in this pr, we have a kag to track this.

**Review by bungle — COMMENTED** [review]
Question.

File: kong/clustering/services/sync/rpc.lua [hunk 6]
```diff
@@ -88,7 +88,7 @@ function _M:init_cp(manager)
       sync_status = CLUSTERING_SYNC_STATUS.NORMAL,
       config_hash = fmt("%032d", default_namespace_version),
       rpc_capabilities = rpc_peers and rpc_peers[node_id] or {},
-    }, { ttl = purge_delay, no_broadcast_crud_event = true })
+    }, { ttl = purge_delay, })
```
Comment (bungle) [inline kong/clustering/services/sync/rpc.lua:91]: Was this causing issues (having `no_broadcast_crud_event = true`, and if so what)?

I think we should at much as possible to add it there. Imagine connecting 1000 DPs and they all send pings or whatever all the time. This table is updated all the time. If we also cause worker events and clustering events it is quite a bad.

**Review by bungle — COMMENTED** [review]
Question.

File: kong/clustering/services/sync/hooks.lua (hunk 1 above)
Comment (bungle) [reply kong/clustering/services/sync/hooks.lua:74]: @StarlightIbuki we need to be careful with debug logs. Currently we don't have means of disabling just some debug logs or fine tune what is debug logged. While log messages like these may be beneficial when debugging incremental sync, it makes debugging everything else much harder as logs are just flooded with these.

File: kong/clustering/services/sync/rpc.lua (hunk 3 above)
Comment (bungle) [reply kong/clustering/services/sync/rpc.lua]: If it didn't cause issues. I would try to remove it from the start.

**Review by StarlightIbuki — APPROVED** [review]

File: kong/db/declarative/export.lua [hunk 7]
```diff
@@ -119,16 +119,9 @@ local function export_from_db_impl(emitter, skip_ws, skip_disabled_entities, exp
 
   local sync_version
   if emitter.want_sync_version then
-    ok, err = db.connector:query("SELECT max(version) from clustering_sync_version", "read")
-    if not ok then
-      return nil, err
-    end
+    local strategy = require("kong.clustering.services.sync.strategies.postgres").new(db)
```
Comment (bungle) [inline kong/db/declarative/export.lua:122]: Can we initialize this once and then just reuse, e.g. a module level `get_latest_version`:

```suggestion
    if not get_latest_version then
      local strategy = require("kong.clustering.services.sync.strategies.postgres").new(db)
      get_latest_version = function()
        return strategy:get_latest_version()
      end
```

Or put it in some helper as it seems to be used in many places.

**Review by bungle — CHANGES_REQUESTED** [review]

File: kong/db/declarative/export.lua (hunk 7 above)
Comment (chronolaw) [reply kong/db/declarative/export.lua:122]: It needs the global variable `kong.db`, I think that it can not be a module level function.

Comment (bungle) [reply kong/db/declarative/export.lua:122]: module level function that caches it on first use

Comment (chronolaw) [reply kong/db/declarative/export.lua:122]: `"kong.clustering.services.sync.strategies.postgres.new()` needs a `db` param, if it is a module level function, `kong.db` may not be initialized.

File: kong/clustering/services/sync/rpc.lua (hunk 3 above)
Comment (chronolaw) [reply kong/clustering/services/sync/rpc.lua]: I will add it in another, both full sync and incremental sync, do you think it is OK?

File: kong/clustering/services/sync/rpc.lua (hunk 6 above)
Comment (chronolaw) [reply kong/clustering/services/sync/rpc.lua:91]: I will add it in another PR, both full sync and incremental sync, do you think it is OK?

File: kong/db/declarative/export.lua (hunk 7 above)
Comment (bungle) [reply kong/db/declarative/export.lua:122]: @chronolaw I meant somethign like this:
```lua
local function get_latest_version do
  local strategy
  local function get_latest_version_real()
    return strategy:get_latest_version()
  end 
  local function get_latest_version()
    strategy = require("kong.clustering.services.sync.strategies.postgres").new(kong.db)
    get_latest_version = get_latest_version_real
    return get_latest_version()
  end
end
```

Comment (chronolaw) [reply kong/db/declarative/export.lua]: done, many thanks.

**Review by bungle — APPROVED** [review]
